src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndexOutOfBounds(index);
+        Node<T> current = getNodeAtIndex(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        linkBefore(newNode, current);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> node = getNodeAtIndex(index);
+        return node.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeAtIndex(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNodeAtIndex(index);
+        unlink(node);
+        size--;
```

Comment:
`size--; `should be inside **unlink** method

Reply:
I tried to put "size--" into unlink method and I received that errors. Tests 3 failed
![2024-01-03_LinkedList_Error_Khrestina](https://github.com/mate-academy/jv-linked-list/assets/136891208/5a67ab99-e001-4c36-9947-b3365e087d81)